services/events/ingestion.go

```diff
@@ -0,0 +1,192 @@
+	return &EventIngestionEngine{
+		subscriber:   subscriber,
+		blocks:       blocks,
+		receipts:     receipts,
+		transactions: transactions,
+		logs:         logs,
+		status:       models.NewEngineStatus(),
+	}
+}
+
+// Ready signals when the engine has started.
+func (e *EventIngestionEngine) Ready() <-chan struct{} {
+	return e.status.IsReady()
+}
+
+// Done signals when the engine has stopped.
+func (e *EventIngestionEngine) Done() <-chan struct{} {
+	// return e.status.IsDone()
+	return nil
+}
+
+// Stop the engine.
+func (e *EventIngestionEngine) Stop() {
+	// todo
+}
+
+// Start the event ingestion engine. Load the latest height that was stored and provide it
+// to the event subscribers as a starting point.
+// Consume the events provided by the event subscriber.
+// Each event is then processed by the event processing methods.
+func (e *EventIngestionEngine) Start(ctx context.Context) error {
+	latest, err := e.blocks.LatestHeight()
+	if err != nil {
+		return err
+	}
+
+	// only init latest height if not set
+	if e.lastHeight == nil {
+		e.lastHeight = models.NewSequentialHeight(latest)
+	} else { // otherwise make sure the latest height is same as the one set on the engine
+		err = e.lastHeight.Increment(latest)
+		if err != nil {
+			return err
+		}
+	}
+
+	e.logs.Info().Uint64("start height", latest).Msg("starting ingestion")
+
+	events, errs, err := e.subscriber.Subscribe(ctx, latest)
+	if err != nil {
+		return err
+	}
+
+	e.status.Ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			e.logs.Info().Msg("event ingestion received done signal")
+			return nil
+
+		case blockEvents, ok := <-events:
+			if !ok {
+				if ctx.Err() != nil {
+					return ctx.Err()
+				}
+				return ErrDisconnected
+			}
+
+			err = e.processEvents(blockEvents)
+			if err != nil {
+				return err
+			}
+
+		case err, ok := <-errs:
+			if !ok {
+				if ctx.Err() != nil {
+					return ctx.Err()
+				}
+
+				return ErrDisconnected
+			}
+
+			return err
+		}
+	}
+}
+
+// processEvents iterates all the events and decides based on the type how to process them.
+func (e *EventIngestionEngine) processEvents(events flow.BlockEvents) error {
+
+	for _, event := range events.Events {
+		if models.IsBlockExecutedEvent(event.Value) {
+			err := e.processBlockEvent(event.Value)
+			if err != nil {
+				return err
+			}
+		} else if models.IsTransactionExecutedEvent(event.Value) {
+			err := e.processTransactionEvent(event.Value)
+			if err != nil {
+				return err
+			}
+		} else {
+			return fmt.Errorf("invalid event type") // should never happen
+		}
+	}
+
+	return nil
+}
+
+func (e *EventIngestionEngine) processBlockEvent(event cadence.Event) error {
+	block, err := models.DecodeBlock(event)
+	if err != nil {
+		return err
+	}
+
+	e.logs.Info().
+		Uint64("height", block.Height).
+		Str("parent hash", block.ParentBlockHash.String()).
+		Msg("ingesting new block executed event")
+
+	if err = e.lastHeight.Increment(block.Height); err != nil {
```

Comment:
The indexing is idempotent meaning even if something goes wrong after updating the latest height, it will be restarted with the height already indexed and re-index that height. But I don't see a reason why wouldn't this be the last action anyway so I will move it.

Reply:
Actually no. After moving I realised the reason that is better to be updated before is to safeguard against any invalid out of sequence heights. So we shouldn't update blocks if height is not incremented by 1 or equal.